docs/proposals/20190704-AutomatedUpgrades.md

```diff
@@ -0,0 +1,196 @@
+The workflow will look roughly like the following:
+
+- an upgrade request is received by `osd` and proxied to `machined`
+- `machined` stops `system` and `k8s.io` services
+- `machined` unmounts all mounts in `/var`
+- `machined` runs the specified installer container
+
+### Automated Upgrades
+
+All of the aformentioned improvements help in the higher goal of automated upgrades.
+Once the above is implemented, upgrades should be simple and self-contained to a machine.
+By running upgrades in containers, we can have very simple operator implementation that can rollout an upgrade across a cluster by simply:
+
+- checking for a new tag
+- orchestrating cluster-wide rollout
+
+The operator will initially coordinate upgrades only, but it will also provide a basis for a number of future automation opportunities.
+
+#### Release Channels
+
+We will maintain a service that allows for the operator to discover when new releases of Talos are made available.
+There will be three "channels" that a machine can subscribe to:
+
+- alpha
+- beta
+- stable
+
+The implementation details around this service and its' interaction with the operator is not in the scope of this proposal.
+
+#### Workflow
+
+In order to upgrade nodes, the operator will subscribe to events for nodes.
+It will keep an in-memory list of all nodes.
+
+The operator will periodically query the target registry for a list of containers within a time range.
+The time range should start where a previous query has ended.
+The initial query should use the currently installed image's timestamp to decide where to start.
+
+In order to supply enough information to the operator, such that it can make an informed decision about using a particular container, we will make use of labels.
+With labels we can communicate to the operator that a container is:
+
+1. an official release
+2. within a given channel
+3. contains a version that is compatible with the current cluster
+
+There may be more information we end up needing, the above is just an example.
+
+Once a container has been chosen, the operator will then create a schedule based on polices.
+The initial policies will perform an upgrade based on:
+
+1. x% of nodes at a given time
+2. availability of a new image in a given channel
+3. time of day
+4. machines per unit of time (e.g. one machine a day, two machines per hour, etc.)
+5. strategy:
+    - In-place
+    - Replace
+
+A user should be able to apply more than one policy.
+
+Once the operator has decided scheduling, it will use the upgrade RPC to perform an upgrades according to the calculated upgrade schedule.
+
+## Rationale
+
+One of the goals in Talos is to allow for machines to be thought of as a container in that it is ephemeral, reproducable, and predictable.
+Since the `kubelet` can reproduce a machine's workload, it is not too crazy to think that we can facilitate upgrades by completely wiping a node, and then performing a fresh install as if the machine is new.
```

Comment:
...with special handling for master (etcd) nodes

Reply:
Glad you said something. Today we perform upgrades of masters by wiping the Etcd data, with the idea that etcd _should_ be able to rebuild itself. The plan was to keep this. and add extra checks when upgrading masters that would ensure etcd is caught up before moving on and upgrading the next master. This could qualify as the "special handling" but curious to get your thoughts on this. The alternative is that we perform a snapshot and restore the snapshot.

Reply:
I'm not fundamentally opposed to the idea, but it is definitely something to be _sure_ to get right.  Regardless of other cluster-wide upgrade settings for node-down tolerance, for example, only one master should be down at any given time and _ideally_ no worker node at the same time (to reduce the change load on the etcd cluster).

I think snapshotting and restoring from snapshot is actually more fraught.  If any of the identifying information for the master node is changed, you can end up with some... not split brain, exactly, but non-agreeing information about the way to contact that peer.  It's better (IMO) to remove the node from membership and then re-add it.

In an ideal world, we would bring up a worker node as a non-voting (learner) etcd member, wait for it to be synchronized, remove-add membership, wait for the master to upgrade, set it to non-voting/learning, then remove-add membership again.

This is a feature of etcd 3.4 specifically to streamline membership changes... however, involving a worker node might result in any number of potential complications, so I don't know that that will ultimately be a _good idea.

Minimally, I think we should make a snapshot before the each node upgrade just in case, but not use that backup unless catastrophe strikes.